alacritty_terminal/src/term/text_run.rs

```diff
@@ -0,0 +1,309 @@
+    /// Opacity of background color of text run content.
+    pub bg_alpha: f32,
+    /// Attributes of this text run (e.g. HIDDEN, STRIKETHROUGH, etc.)
+    pub flags: Flags,
+}
+
+impl TextRun {
+    // These two constructors are used by TextRunIter and are not widely applicable
+    fn from_iter_state(
+        start: RunStart,
+        (latest, is_wide): LatestCol,
+        buffer: (String, Vec<[char; MAX_ZEROWIDTH_CHARS]>),
+    ) -> Self {
+        let end_column = if is_wide { latest + 1 } else { latest };
+        TextRun {
+            line: start.line,
+            span: (start.column, end_column),
+            content: TextRunContent::CharRun(buffer.0, buffer.1),
+            fg: start.fg,
+            bg: start.bg,
+            bg_alpha: start.bg_alpha,
+            flags: start.flags,
+        }
+    }
+
+    fn from_cursor_rc(start: RunStart, cursor: CursorKey) -> Self {
+        TextRun {
+            line: start.line,
+            span: (start.column, start.column),
+            content: TextRunContent::Cursor(cursor),
+            fg: start.fg,
+            bg: start.bg,
+            bg_alpha: start.bg_alpha,
+            flags: start.flags,
+        }
+    }
+
+    /// Holdover method while converting from rendering Cells to TextRuns
+    fn cell_at(&self, col: Column) -> RenderableCell {
+        RenderableCell {
+            line: self.line,
+            column: col,
+            inner: RenderableCellContent::Chars([' '; crate::term::cell::MAX_ZEROWIDTH_CHARS + 1]),
+            fg: self.fg,
+            bg: self.bg,
+            bg_alpha: self.bg_alpha,
+            flags: self.flags,
+        }
+    }
+
+    /// Number of columns this TextRun spans
+    pub fn len(&self) -> usize {
+        let (start, end) = self.span;
+        end.0 - start.0
+    }
+
+    /// True if TextRun contains characters, false if it contains no characters.
+    pub fn is_empty(&self) -> bool {
+        match &self.content {
+            TextRunContent::Cursor(_) => false,
+            TextRunContent::CharRun(ref string, ref zero_widths) => {
+                string.is_empty() && zero_widths.is_empty()
+            },
+        }
+    }
+
+    /// First column of the run
+    pub fn start_col(&self) -> Column {
+        self.span.0
+    }
+
+    /// First cell in the TextRun
+    pub fn start_cell(&self) -> RenderableCell {
+        self.cell_at(self.span.0)
+    }
+
+    /// End cell in the TextRun
+    pub fn end_cell(&self) -> RenderableCell {
+        self.cell_at(self.span.1)
+    }
+
+    /// First point covered by this TextRun
+    pub fn start_point(&self) -> Point {
+        Point { line: self.line, col: self.span.0 }
+    }
+
+    /// End point covered by this TextRun
+    pub fn end_point(&self) -> Point {
+        Point { line: self.line, col: self.span.1 }
+    }
+
+    /// Returns iterator over range of columns [run.0, run.1]
+    pub fn cols(&self) -> impl Iterator<Item = Column> {
+        let (start, end) = self.span;
```

Comment:
We can take destructuring even further here!

```rust
let (Column(start), Column(end)) = self.span;
```

This way there's no need to do `.0` anymore.

I'd also move this right above the call itself, just since this is a very temporary variable definition just for ergonomics.